video_xblock/mixins.py

```diff
@@ -165,86 +195,98 @@ def convert_3playmedia_caps_to_vtt(self, caps, video_id, lang="en", lang_label="
-                "label": lang_label,
-                "url": external_url,
-                "source": source,
-            })
-        log.debug("Got 3PlayMedia transcripts: {}".format(translations))
-        return 'success', translations
+        for transcript_data in transcripts_list:
+            transcript = self.fetch_single_3pm_translation(transcript_data)
+            if transcript is None:
+                raise StopIteration
+            transcript_ordered_dict = transcript._asdict()
+            transcript_ordered_dict['content'] = ''  # we don't want to parse it to JSON
+            yield transcript_ordered_dict
 
-    @XBlock.json_handler
-    def get_transcripts_3playmedia_api_handler(self, data, _suffix=''):
+    def get_3pm_transcripts_list(self, file_id, apikey):
         """
-        Xblock handler to authenticate to a video platform's API. Called by JavaScript of `studio_view`.
+        Make API request to fetch list of available transcripts for given file ID.
 
-        Arguments:
-            data (dict): Data from frontend, necessary for authentication (tokens, account id, etc).
-            suffix (str): Slug used for routing.
-        Returns:
-            response (dict): Status messages key-value pairs.
+        :return: (list of dicts OR dict) all available transcripts attached to file with ID OR error dict
         """
-        apikey = data.get('api_key', self.threeplaymedia_apikey) or ''
-        file_id = data.get('file_id', '')
-        status, transcripts = self.get_translations_from_3playmedia(
-            apikey=apikey, file_id=file_id
-        )
-        if status == 'error':
-            return transcripts
+        domain = self.THREE_PLAY_MEDIA_API_DOMAIN
+
+        transcripts_list = []
+        failure_message = _("3PlayMedia transcripts fetching API request has failed!")
+        success_message = _("3PlayMedia transcripts fetched successfully.")
+        feedback = {'status': Status.error, 'message': failure_message}
 
-        return {
-            'transcripts': transcripts,
-            'success_message': _(
-                'Successfully fetched transcripts from 3playMedia. Please check transcripts list above.'
+        try:
+            response = requests.get(
+                '{domain}files/{file_id}/transcripts?apikey={api_key}'.format(
+                    domain=domain, file_id=file_id, api_key=apikey
+                )
             )
-        }
+        except IOError:
+            log.exception(failure_message)
+            return feedback, transcripts_list
+
+        if response.ok and isinstance(response.json(), list):
+            transcripts_list = response.json()
+            feedback['status'] = Status.success
+            feedback['message'] = success_message
+        else:
+            feedback['status'] = Status.error
+        return feedback, transcripts_list
+
+    def fetch_single_3pm_translation(self, transcript_data, format_id=TPMApiTranscriptFormatID.WEBVTT):
```

Comment:
Here we fetch up to date transcript from 3PlayMedia and serve it to the video player.